libc-bottom-half/crt/crt1.c

```diff
@@ -13,29 +13,47 @@ void _Exit(int) __attribute__((noreturn));
 static __wasi_errno_t populate_args(size_t *argc, char ***argv) {
     __wasi_errno_t err;
 
-    /* Get the sizes of the arrays we'll have to create to copy in the args. */
+    // Get the sizes of the arrays we'll have to create to copy in the args.
     size_t argv_buf_size;
-    err = __wasi_args_sizes_get(argc, &argv_buf_size);
+    size_t new_argc;
+    err = __wasi_args_sizes_get(&new_argc, &argv_buf_size);
     if (err != __WASI_ESUCCESS) {
         return err;
     }
-    if (*argc == 0) {
+    if (new_argc == 0) {
         return __WASI_ESUCCESS;
     }
 
-    /* Allocate memory for the array of pointers, adding null terminator. */
-    *argv = malloc(sizeof(char *) * (*argc + 1));
-    /* Allocate memory for storing the argument chars. */
-    char *argv_buf = malloc(sizeof(char) * argv_buf_size);
-    if (*argv == NULL || argv_buf == NULL) {
+    // Add 1 for the NULL pointer to mark the end, and check for overflow.
+    size_t num_ptrs = new_argc + 1;
+    if (num_ptrs == 0) {
         return __WASI_ENOMEM;
     }
 
-    /* Make sure the last pointer in the array is NULL. */
-    *argv[*argc] = NULL;
+    // Allocate memory for storing the argument chars.
+    char *argv_buf = malloc(argv_buf_size);
```

Comment:
In emscripten I ended up inlining this function into _start itself so that these could become `alloca` and simple programs could avoid depending malloc + free.  Not related to this change but worth mentioning for later maybe?

Reply:
That's a good idea. Once the libpreopen reorg PR lands, and the next PR to make it so that the libpreopen code doesn't get linked in if you don't use it, then this will be the only `malloc` left, so we should remove it. I'll submit a separate PR for that.